### PR TITLE
TFLu: Port TFL detection postprocess operator

### PR DESCRIPTION
**Pull request by mansnils**
This is fixing issue: https://github.com/tensorflow/tensorflow/issues/43565

### CODE REVIEW

File: tensorflow/lite/micro/test_helpers.cc [hunk 1]
```diff
@@ -884,7 +884,11 @@ TfLiteTensor CreateFloatTensor(const float* data, TfLiteIntArray* dims,
   TfLiteTensor result = CreateTensor(dims, is_variable);
   result.type = kTfLiteFloat32;
   result.data.f = const_cast<float*>(data);
-  result.bytes = ElementCount(*dims) * sizeof(float);
```
[the remark this answers is not in the document]
Comment (mansnils) [reply tensorflow/lite/micro/test_helpers.cc:887]: You are right, it is not needed. Removing it.

File: tensorflow/lite/micro/tools/make/third_party_downloads.inc [hunk 2]
```diff
@@ -8,13 +8,8 @@
...
 GEMMLOWP_MD5 := "7e8191b24853d75de2af87622ad293ba"
 
-ifeq ($(HOST_OS),windows)
-  FLATBUFFERS_URL := "http://mirror.tensorflow.org/github.com/google/flatbuffers/archive/v1.12.0.zip"
-  FLATBUFFERS_MD5 := "a1afdbf114dec01a861c1b8c917d0fc7"
-else
-  FLATBUFFERS_URL := "http://mirror.tensorflow.org/github.com/google/flatbuffers/archive/v1.12.0.tar.gz"
-  FLATBUFFERS_MD5 := "c62ffefb3d4548b127cca14ce047f16c"
-endif
+FLATBUFFERS_URL := "https://github.com/google/flatbuffers/archive/dca12522a9f9e37f126ab925fd385c807ab4f84e.zip"
```
[the remark this answers is not in the document]
Comment (mansnils) [reply tensorflow/lite/micro/tools/make/third_party_downloads.inc:11]: I only tested it on Linux. So I am proposing to revert the change in this file and suppress the error in the operator with a comment about this.

**Comment by gbaned** [comment]
@mansnils  Can you please resolve conflicts? Thanks!

**Comment by gbaned** [comment]
@mansnils Any update on this PR? Please. Thanks!

**Comment by mansnils** [comment]
Removing support for unknown output dimensions. Because it does not work with TFLiteEvalTensor API and it would require major changes in TFLu.
Most models seen with this operator have unknown output dimensions. So to run those they have to be modified first so that the out dimensions are known beforehand.

**Review by njeffrie — APPROVED** [review]

**Review by advaitjain — CHANGES_REQUESTED** [review]
apologies for the delayed review here. Mostly Google-style nits.

I have also triggered the CI build so you can see the CI failure log.

File: tensorflow/lite/micro/all_ops_resolver.cc [hunk 3]
```diff
@@ -84,6 +85,8 @@ AllOpsResolver::AllOpsResolver() {
   AddUnpack();
 
   // TODO(b/159644355): Figure out if custom Ops belong in AllOpsResolver.
+  AddCustom("TFLite_Detection_PostProcess",
+            tflite::ops::micro::custom::Register_DETECTION_POSTPROCESS());
```
[the remark this answers is not in the document]
Comment (mansnils) [reply tensorflow/lite/micro/all_ops_resolver.cc:89]: All right

File: tensorflow/lite/micro/kernels/detection_postprocess_test/readme [hunk 4]
```diff
@@ -0,0 +1,22 @@
+# Copyright 2020 The TensorFlow Authors. All Rights Reserved.
+#
```
[the remark this answers is not in the document]
Comment (mansnils) [reply tensorflow/lite/micro/kernels/detection_postprocess_test/readme:2]: ok

**Review by advaitjain — APPROVED** [review]

**Comment by rthadur** [comment]
@mansnils can you please check sanity build failures ?

**Comment by advaitjain** [comment]
A few of the CI failures:
 * the x86 build is failing for the TfLite Micro CI build. I think you need the arrays to be unsigned char instead of char.
 * the other two (MacOS CPU Python3 and Ubuntu Sanity) are related to the bazel build and BUILD file formatting.

I think you can recreate with:
```
bazel test tensorflow/lite/micro:micro_interpreter_test
```

And buildifier can be installed using instructions from [here](https://github.com/bazelbuild/buildtools/tree/master/buildifier).

I can help tomorrow if you have trouble reproducing these errors locally.

**Comment by advaitjain** [comment]
> 
> ```
> bazel test tensorflow/lite/micro:micro_interpreter_test
> ```
> 

This one likely needs the arena size to be updated.

**Comment by mansnils** [comment]
@advaitjain Thanks for the tip about: bazel test tensorflow/lite/micro:micro_interpreter_test. It turns out that it TestIncompleteInitializationAllocationsWithSmallArena that fails in this line:
   TF_LITE_MICRO_EXPECT_EQ(
      static_cast<size_t>(128),
       allocator->GetSimpleMemoryAllocator()->GetHeadUsedBytes()); And it turns out this is caused by the change of kMaxScratchBuffersPerOp  in micro_allocator.cc. It seems allocator->GetSimpleMemoryAllocator()->GetHeadUsedBytes() will change depending on kMaxScratchBuffersPerOp in this test.

**Comment by mansnils** [comment]
I am pushing the fix for the micro_interpreter_test to see what tests it is fixing

**Comment by advaitjain** [comment]
@mansnils, I have pushed a commit fixing some remaining issues with https://github.com/tensorflow/tensorflow/pull/43566/commits/5491648d7f710d445ffe873afbec0b9c7834be47.

This PR should be ready to go now.

**Review by advaitjain — APPROVED** [review]

**Comment by advaitjain** [comment]
Unfortunately, the TFLM build is still failing. I think we're going to have to take a second look at the kernel runner and see how we can fix this.

I'll take a look tomorrow.

**Comment by mansnils** [comment]
There is error: narrowing conversion of '-108' from 'int' to 'char' inside { } [-Wnarrowing] because of
const char g_gen_data_none_regular_nms[] = { .. -108
Normally that would be wrong but in this case it is actually intentional so I figure it is ok to use
#pragma GCC diagnostic ignored "-Wnarrowing"
in the generated code.

**Comment by mansnils** [comment]
@advaitjain  To get consistent results between TFL and TFLM I am considering changing this in both kernels.
From:
    float ycenter = box_centersize.y / scale_values.y * anchor.h + anchor.y;
    float xcenter = box_centersize.x / scale_values.x * anchor.w + anchor.x;
    float half_h =
        0.5f * static_cast<float>(std::exp(box_centersize.h / scale_values.h)) *
        anchor.h;
    float half_w =
        0.5f * static_cast<float>(std::exp(box_centersize.w / scale_values.w)) *
 anchor.w;
To:
    float ycenter = static_cast<float>(
        static_cast<double>(box_centersize.y) /
        static_cast<double>(scale_values.y) *
        static_cast<double>(anchor.h) + static_cast<double>(anchor.y));

    float xcenter =  static_cast<float>(
        static_cast<double>(box_centersize.x) /
        static_cast<double>(scale_values.x) *
        static_cast<double>(anchor.w) + static_cast<double>(anchor.x));

    float half_h = static_cast<float>(
        0.5 * (std::exp(static_cast<double>(box_centersize.h) /
                        static_cast<double>(scale_values.h))) *
        static_cast<double>(anchor.h));
    float half_w =  static_cast<float>(
        0.5 * (std::exp(static_cast<double>(box_centersize.w) /
                        static_cast<double>(scale_values.w))) *
                        static_cast<double>(anchor.w));
Question: Would it it be ok to change this in the TFL kernel or could it break some test?

**Comment by mansnils** [comment]
Sorry, forget the comment about: #pragma GCC diagnostic ignored "-Wnarrowing"

**Comment by mansnils** [comment]
@advaitjain  To get rid of -Wnarrowing we should generate const int8_t* and cast back to const char* in detection_postprocess_test.cc. I was about to push this (after downloading your last commit) but have some problem:

error: failed to push some refs to 'git@github.com:mansnils/tensorflow.git'
hint: Updates were rejected because the remote contains work that you do
hint: not have locally. This is usually caused by another repository pushing
hint: to the same ref. You may want to first integrate the remote changes
hint: (e.g., 'git pull ...') before pushing again.
hint: See the 'Note about fast-forwards' in 'git push --help' for details.

How do I pull your changes?

**Comment by advaitjain** [comment]
> @advaitjain To get rid of -Wnarrowing we should generate const int8_t* and cast back to const char* in detection_postprocess_test.cc. I was about to push this (after downloading your last commit) but have some problem:
> 
> error: failed to push some refs to '[git@github.com](mailto:git@github.com):mansnils/tensorflow.git'
> hint: Updates were rejected because the remote contains work that you do
> hint: not have locally. This is usually caused by another repository pushing
> hint: to the same ref. You may want to first integrate the remote changes
> hint: (e.g., 'git pull ...') before pushing again.
> hint: See the 'Note about fast-forwards' in 'git push --help' for details.
> 
> How do I pull your changes?

I think if you git pull that would do the trick. I had pushed to your remote branch so you local branch was a couple commits behind.

I have made the fix that you suggested with https://github.com/tensorflow/tensorflow/pull/43566/commits/aad242613d71dbc38bc7a8a0633e31fb192ebb85 so this issue is resolved.

**Comment by advaitjain** [comment]
> @advaitjain To get consistent results between TFL and TFLM I am considering changing this in both kernels.
> From:
> 
> ```
> float ycenter = box_centersize.y / scale_values.y * anchor.h + anchor.y;
> float xcenter = box_centersize.x / scale_values.x * anchor.w + anchor.x;
> float half_h =
>     0.5f * static_cast<float>(std::exp(box_centersize.h / scale_values.h)) *
>     anchor.h;
> float half_w =
>     0.5f * static_cast<float>(std::exp(box_centersize.w / scale_values.w)) *  anchor.w;
> ```
> 
> To:
> 
> ```
> float ycenter = static_cast<float>(
>     static_cast<double>(box_centersize.y) /
>     static_cast<double>(scale_values.y) *
>     static_cast<double>(anchor.h) + static_cast<double>(anchor.y));
> float xcenter =  static_cast<float>(
>     static_cast<double>(box_centersize.x) /
>     static_cast<double>(scale_values.x) *
>     static_cast<double>(anchor.w) + static_cast<double>(anchor.x));
> float half_h = static_cast<float>(
>     0.5 * (std::exp(static_cast<double>(box_centersize.h) /
>                     static_cast<double>(scale_values.h))) *
>     static_cast<double>(anchor.h));
> float half_w =  static_cast<float>(
>     0.5 * (std::exp(static_cast<double>(box_centersize.w) /
>                     static_cast<double>(scale_values.w))) *
>                     static_cast<double>(anchor.w));
> ```
> 
> Question: Would it it be ok to change this in the TFL kernel or could it break some test?

Being explicit about float vs double math seems very reasonable to me. It would be nice if we could make this less verbose though. I'd be surprised if it breaks something.

Let's keep this change for a follow-on PR.

**Comment by advaitjain** [comment]
Summarizing where we are with this PR for @mansnils @njeffrie and me:

 * the flex buffer generation and char vs unsinged char is resolved and all tests pass with both bazel and make (on x86)
 * build files etc are all formatted

We are still back to a different version of the original problem on bluepill:
```
make -f tensorflow/lite/micro/tools/make/Makefile TARGET=bluepill kernel_detection_postprocess_test
```

gives the following error:
```
`__lock___malloc_recursive_mutex' referenced in section `.text.__malloc_lock' of /home/advaitjain/github/tensorflow/tensorflow/lite/micro/tools/make/downloads/gcc_embedded/bin/../lib/gcc/arm-none-eabi/7.3.1/../../../../arm-none-eabi/lib/thumb/v7-m/libc.a(lib_a-mlock.o): defined in discarded section `COMMON' of /home/advaitjain/github/tensorflow/tensorflow/lite/micro/tools/make/downloads/gcc_embedded/bin/../lib/gcc/arm-none-eabi/7.3.1/../../../../arm-none-eabi/lib/thumb/v7-m/libc.a(lib_a-lock.o)
`__lock___malloc_recursive_mutex' referenced in section `.text.__malloc_unlock' of /home/advaitjain/github/tensorflow/tensorflow/lite/micro/tools/make/downloads/gcc_embedded/bin/../lib/gcc/arm-none-eabi/7.3.1/../../../../arm-none-eabi/lib/thumb/v7-m/libc.a(lib_a-mlock.o): defined in discarded section `COMMON' of /home/advaitjain/github/tensorflow/tensorflow/lite/micro/tools/make/downloads/gcc_embedded/bin/../lib/gcc/arm-none-eabi/7.3.1/../../../../arm-none-eabi/lib/thumb/v7-m/libc.a(lib_a-lock.o)
`errno' referenced in section `.text._sbrk_r' of /home/advaitjain/github/tensorflow/tensorflow/lite/micro/tools/make/downloads/gcc_embedded/bin/../lib/gcc/arm-none-eabi/7.3.1/../../../../arm-none-eabi/lib/thumb/v7-m/libc.a(lib_a-sbrkr.o): defined in discarded section `COMMON' of /home/advaitjain/github/tensorflow/tensorflow/lite/micro/tools/make/downloads/gcc_embedded/bin/../lib/gcc/arm-none-eabi/7.3.1/../../../../arm-none-eabi/lib/thumb/v7-m/libc.a(lib_a-reent.o)
/home/advaitjain/github/tensorflow/tensorflow/lite/micro/tools/make/downloads/gcc_embedded/bin/../lib/gcc/arm-none-eabi/7.3.1/../../../../arm-none-eabi/lib/thumb/v7-m/libc.a(lib_a-sbrkr.o): In function `_sbrk_r':
sbrkr.c:(.text._sbrk_r+0xc): undefined reference to `_sbrk'
```

The error message is puzzling to me since it is saying something about being in the discarded section which makes me wonder why its an issue.

@mansnils: you should be able to `git pull` and have my changes appear on your local branch as well. If you can debug this _sbrk issue, that would be great.

**Comment by mansnils** [comment]
It turns out the _sbrk issue is caused by the called to Flexbuffers AsFloat() in PPD. That call will reference  flatbuffers::StringToNumber(), which is the culprit. flatbuffers::StringToNumber() is not called directly so it can be removed with the unit test for PPD still passing. I am not sure how to resolve this issue in the best way. However I am proposing a temporary workaround to patch flexbuffers.h for Bluepill.

**Comment by advaitjain** [comment]
Agreed that we should patch flatbuffers as a workaround for now.

I'd prefer having the patching be centralized instead of in the bluepill makefile.

I'm going to do this in two steps:
 1. https://github.com/tensorflow/tensorflow/pull/44860 pulls the flatbuffer downloads into its own script.
 2. I'll add a follow-on PR that does the patching. I have some prototyping commands here and will put them into a PR on Monday:
```bash
string_to_num_line=`awk '/StringToNumber/{ print NR; }' flexbuffers.h`
case_string_line=$((${string_to_num_line} - 2))

head -n ${case_string_line} flexbuffers.h > new_flexbuffers.h

echo "#if 1" >> new_flexbuffers.h
echo "          // Introduce a segfault for an unsupported code path for TFLM." >> new_flexbuffers.h
echo "          return *(static_cast<double*>(nullptr));" >> new_flexbuffers.h
echo "#else" >> new_flexbuffers.h
echo "          // This is the original code" >> new_flexbuffers.h
sed -n -e $((${string_to_num_line} -  1)),$((${string_to_num_line} + 1))p flexbuffers.h >> new_flexbuffers.h
echo "#endif" >> new_flexbuffers.h

total_num_lines=`wc -l flexbuffers.h | awk '{print $1}'`
sed -n -e $((${string_to_num_line} + 2)),${total_num_lines}p flexbuffers.h >> new_flexbuffers.h
```

File: tensorflow/lite/micro/kernels/test_data_generation/generate_flexbuffers_data.cc [hunk 5]
```diff
@@ -52,18 +58,28 @@ void generate(const char* name, bool use_regular_nms) {
...
-    printf("0x%02x, ", init_data[i]);
+    printf("0x%02x,", init_data[i]);
+    count_newlines++;
+    if ((count_newlines % 12) == 0) {
+      printf("\n");
+      print_spaces(4);
+    }
+    else if (i < (fbb_size-1)) {
+      print_spaces(1);
+    }
```
Comment (advaitjain) [inline tensorflow/lite/micro/kernels/test_data_generation/generate_flexbuffers_data.cc:73]: I think we can skip this formatting code. Its easier to clang-format the output file.

**Comment by advaitjain** [comment]
Created #44915 to patch flatbuffers to avoid pulling in strtod. Once that is merged, we should be able to revert 18ec320.

@mansnils, I'd suggest removing 18ec320 and then force-push.

**Comment by mansnils** [comment]
@advaitjain Ok, will do. Will just correct the name of the readme file.

**Review by advaitjain — APPROVED** [review]

**Comment by gbaned** [comment]
@mansnils  Can you please address Ubuntu Sanity errors? Thanks!

**Comment by mansnils** [comment]
@gbaned Ubuntu Sanity error fixed

**Review by advaitjain — APPROVED** [review]

**Review by advaitjain — APPROVED** [review]

**Comment by advaitjain** [comment]
Pushed a commit to fix internal errors. Let's see how the merging goes this time.

**Comment by mansnils** [comment]
@advaitjain After pulling your changes, ci_sanity.sh no longer passed so I pushed https://github.com/tensorflow/tensorflow/pull/43566/commits/0bc6594e93d9f1326d1e359f274ab531b2da4730. At least now Ubunty Sanity passed but Windows and Mac builds still failed. Is it stability issues or what has changed?

**Review by advaitjain — APPROVED** [review]

**Comment by advaitjain** [comment]
> @advaitjain After pulling your changes, ci_sanity.sh no longer passed so I pushed [0bc6594](https://github.com/tensorflow/tensorflow/commit/0bc6594e93d9f1326d1e359f274ab531b2da4730). At least now Ubunty Sanity passed but Windows and Mac builds still failed. Is it stability issues or what has changed?

Thanks Mans. The Win and Mac errors look like existing failures. Approved this PR and will see what the internal checks have to say.

**Comment by advaitjain** [comment]
***(sigh)*** Internal checks failed but the external CI for Micro passed due to discrepancy between internal and external.

This PR makes the external CI closer to the internal checks: https://github.com/tensorflow/tensorflow/pull/45018

**Comment by mansnils** [comment]
@advaitjain It seems the iOS error is an existing error, right?

**Review by advaitjain — APPROVED** [review]

**Comment by advaitjain** [comment]
Thanks for the fix. Yes, the iOS error is existing and unrelated.

**Comment by advaitjain** [comment]
Still having issues unfortunately :-/

The CI is failing with:
```
tensorflow/lite/micro/kernels/detection_postprocess.cc: In function ‘TfLiteStatus tflite::{anonymous}::Eval(TfLiteContext*, TfLiteNode*)’:
tensorflow/lite/micro/kernels/detection_postprocess.cc:591:58: error: potential null pointer dereference [-Werror=null-dereference]
   tflite::micro::GetTensorData<float>(num_detections)[0] =
   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^
       size_of_sorted_indices;
       ~~~~~~~~~~~~~~~~~~~~~~                              
tensorflow/lite/micro/kernels/detection_postprocess.cc:687:58: error: potential null pointer dereference [-Werror=null-dereference]
   tflite::micro::GetTensorData<float>(num_detections)[0] = output_box_index;
```

but it is not reproducible locally.

My current guess is that the gcc version used as part of CI is old (gcc-6.3.0). I have a change under review that updates https://github.com/tensorflow/tensorflow/blob/3d3c4aaaedaf117b3b8cc9c30a594889094db6f1/tensorflow/tools/ci_build/Dockerfile.micro#L4

to
```
python:3.9.0-buster
```

which should mean that we use gcc 8.3.0 as part of CI. Let's see what happens when that change get merged.

**Comment by mansnils** [comment]
@advaitjain It seems like the updated docker didn't help, or do we need to rebase?

**Comment by advaitjain** [comment]
> Still having issues unfortunately :-/
> 
> The CI is failing with:
> 
> ```
> tensorflow/lite/micro/kernels/detection_postprocess.cc: In function ‘TfLiteStatus tflite::{anonymous}::Eval(TfLiteContext*, TfLiteNode*)’:
> tensorflow/lite/micro/kernels/detection_postprocess.cc:591:58: error: potential null pointer dereference [-Werror=null-dereference]
>    tflite::micro::GetTensorData<float>(num_detections)[0] =
>    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^
>        size_of_sorted_indices;
>        ~~~~~~~~~~~~~~~~~~~~~~                              
> tensorflow/lite/micro/kernels/detection_postprocess.cc:687:58: error: potential null pointer dereference [-Werror=null-dereference]
>    tflite::micro::GetTensorData<float>(num_detections)[0] = output_box_index;
> ```
> 
> but it is not reproducible locally.
> 
> My current guess is that the gcc version used as part of CI is old (gcc-6.3.0). I have a change under review that updates
> 
> https://github.com/tensorflow/tensorflow/blob/3d3c4aaaedaf117b3b8cc9c30a594889094db6f1/tensorflow/tools/ci_build/Dockerfile.micro#L4
> 
> to
> 
> ```
> python:3.9.0-buster
> ```
> 
> which should mean that we use gcc 8.3.0 as part of CI. Let's see what happens when that change get merged.


I wasn't able to reproduce locally because I hadn't done a `make clean_downloads`.

Created https://github.com/tensorflow/tensorflow/pull/45286 to address this issue. [This comment](https://github.com/tensorflow/tensorflow/issues/44971#issuecomment-736130061) has some additional context.

**Review by advaitjain — APPROVED** [review]

**Pull request merged**